solara/_stores.py

```diff
@@ -0,0 +1,180 @@
+import copy
+import dataclasses
+import inspect
+from typing import Callable, ContextManager, Generic, Optional
+import warnings
+from .toestand import ValueBase, KernelStore, S, _find_outside_solara_frame
+import solara.util
+
+
+@dataclasses.dataclass
+class StoreValue(Generic[S]):
+    private: S  # the internal private value, should never be mutated
+    public: Optional[S]  # this is the value that is exposed in .get(), it is a deep copy of private
+    get_traceback: Optional[inspect.Traceback]
+    set_value: Optional[S]  # the value that was set using .set(..), we deepcopy this to set private
+    set_traceback: Optional[inspect.Traceback]
+
+
+class MutateDetectorStore(ValueBase[S]):
+    def __init__(self, store: KernelStore[StoreValue[S]], equals=solara.util.equals):
+        self._storage = store
+        self._enabled = True
+        super().__init__(equals=equals)
+
+    @property
+    def lock(self):
+        return self._storage.lock
+
+    def get(self) -> S:
+        print("check!")
+        self.check_mutations()
+        self._ensure_public_exists()
+        value = self._storage.get()
+        assert value.public is not None
```

Comment:
In the dataclass we document this:
```
# this is the value that is exposed in .get(), it is a deep copy of private
```

But since we do this lazily, it can be None (_ensure_public_exists should make sure it doesn't, but this is to also get the typing right, and to self document)